src/main/scala/io/iohk/ethereum/blockchain/sync/PeersClient.scala

```diff
@@ -39,6 +43,8 @@ class PeersClient(
   val statusSchedule: Cancellable =
     scheduler.scheduleWithFixedDelay(syncConfig.printStatusInterval, syncConfig.printStatusInterval, self, PrintStatus)
 
+  val activeFetchingNodes: mutable.Set[Peer] = lruSet[Peer](syncConfig.peersToFetchFrom)
```

Comment:
Is the LRU set the right behavior for this and would it be possible to split this out into something that is unit-testable?

Don't we want to reject peers based on something other than "least recently used"? What's the definition of "used" here btw?

Reply:
we are rejecting peers for multiple reasons, but this piece of code has nothing to do to rejection, it's needed for picking best N peers out of our pool of discovered and connected peers, and we need to make sure that when we fetch from N chosen peers in parallel, all the peers are actually different, hence lru set. "used" here in the meaning we are fetching from it at the moment.

If you have an alternative approach for this - I don't mind changing this piece.

Reply:
An immutable set of peers? Then the logic that operate on the set can be unit tested, separate from where it is kept as state.

Reply:
and how to make sure that every Nth peer replaces the oldest one from the set? reimplement lruSet?

Reply:
Yes, and using immutable data structures.